site/public/js/notebook_builder/notebook-builder.js

```diff
@@ -0,0 +1,106 @@
+// Attach notebook_builder object to the window
+window.addEventListener('DOMContentLoaded', (event) => {
+    window.notebook_builder = new NotebookBuilder();
+    notebook_builder.render();
+});
+
+class NotebookBuilder {
+    constructor() {
+        this.widgets = [new SelectorWidget()];
+    }
+
+    /**
+     * Display all widgets the notebook builder is keeping track of
+     */
+    render() {
+        // Get a handle on the main div
+        const main_div = document.getElementById('notebook-builder');
+
+        // Clear
+        main_div.innerHTML = '';
+
+        // Draw widgets
+        this.widgets.forEach(widget => {
+            main_div.appendChild(widget.render());
+        });
+    }
+
+    /**
+     * From the list of widgets compile and return a 'notebook' json
+     */
+    getJSON() {
+        const notebook_array = [];
+
+        // Iterate but dont include the final widget which is the selector widget
+        let i;
+        for (i = 0; i < this.widgets.length - 1; i++) {
+
+            // Ensure we got something back before adding to the notebook_array
+            const widget_json = this.widgets[i].getJSON();
+            if (widget_json) {
+                notebook_array.push(widget_json);
+            }
+        }
+
+        return {
+            notebook: notebook_array
+        };
+    }
+
+    /**
+     * Add a widget to the notebook builder form
+     *
+     * @param widget
+     */
+    widgetAdd(widget) {
+        this.widgets.splice(this.widgets.length - 1, 0, widget);
```

Comment:
Is this not just `this.widgets.push(widget)`?

Reply:
No the widget array also contains the 'selector' widget which should always be last.  Any new widgets should be inserted next to last.

Reply:
I think I may end up refactoring this when it's needed to support a top, mid, and bottom widget arrays.  Bottom will hold the selector.  Top will hold configuration options that are not part of the notebook, for example assignment message.  Then mid will hold notebook widgets and will be the only part that allows reordering.  This will simplify some of the logic for rearrangement of widgets.

Reply:
Can we not do this now with this bottom widget? Unless there's a design that we're going to be swapping the bottom out with something else, making it part of this list is just confusing and makes it too easy to introduce sorting bugs for the unwary.